File: Iverson.Clients/TypeScript/tests/pipeline.test.ts
import { describe, expect, it } from 'vitest';
import { pipeline } from '../src/pipeline.js';
import {
    AggregationType,
    DateTrunc,
    JoinKind,
    SearchOperator,
    WindowFunctionKind,
} from '../generated/object_search.js';

describe('PipelineBuilder', () => {
    it('compiles a full pipeline to the expected proto', () => {
        const request = pipeline('Article')
            .where('IsPublished', SearchOperator.EQUALS, true)
            .step('by_author', s => s
                .groupBy('AuthorId')
                .countAll('articles')
                .having('articles', SearchOperator.GREATER_THAN, 5))
            .step('ranked', s => s
                .rowNumber('rank', { orderBy: 'articles', descending: true }))
            .step('named', s => s
                .join('Author', 'AuthorId', 'Id')
                .select(sel => sel.allFrom('ranked').pick('Author', 'Name', 'author_name')))
            .sortOnDesc('rank')
            .limit(5)
            .build();

        expect(request.typeName).toBe('Article');
        expect(request.baseWhere).toHaveLength(1);
        expect(request.steps).toHaveLength(3);

        const agg = request.steps[0];
        expect(agg.name).toBe('by_author');
        expect(agg.groupBy[0]).toMatchObject({ field: 'AuthorId', dateTrunc: DateTrunc.NONE });
        expect(agg.metrics[0]).toMatchObject({ name: 'articles', type: AggregationType.COUNT });
        expect(agg.having[0].property).toBe('articles');

        const win = request.steps[1];
        expect(win.windows[0]).toMatchObject({
            alias: 'rank', kind: WindowFunctionKind.ROW_NUMBER,
            orderBy: 'articles', descending: true,
        });

        const joined = request.steps[2];
        expect(joined.joins[0]).toMatchObject({ source: 'Author', kind: JoinKind.INNER });
        expect(joined.joins[0].on[0]).toMatchObject({ left: 'AuthorId', right: 'Id' });
        expect(joined.select[0].all).toBe(true);
        expect(joined.select[1].alias).toBe('author_name');

        expect(request.limit).toBe(5);
    });

    it('carries explicit reads and defaults the limit', () => {
        const request = pipeline('Article')
            .step('a', s => s.derive('x', 'WordCount + 1'))
            .step('b', s => s.reads('base').derive('y', 'WordCount + 2'))
            .build();

        expect(request.steps[1].reads).toBe('base');
        expect(request.limit).toBe(10_000);
    });

    it('throws on duplicate step names', () => {
        const b = pipeline('Article').step('x', s => s.derive('a', 'WordCount'));
        expect(() => b.step('X', s => s.derive('b', 'WordCount'))).toThrow(/X/);
    });

    it('throws on reads of an unknown step', () => {
        expect(() => pipeline('Article').step('a', s => s.reads('nope'))).toThrow(/nope/);
    });

    it('throws when windows and groupBy share a step', () => {
        expect(() => pipeline('Article').step('bad', s => s
            .rowNumber('rn', { orderBy: 'Id' })
            .groupBy('AuthorId').countAll('n'))).toThrow(/bad/);
    });

    it('throws on a join without a select', () => {
        expect(() => pipeline('Article').step('bad', s => s.join('Author', 'AuthorId', 'Id')))
            .toThrow(/select/);
    });

    it('throws on duplicate aliases within a step', () => {
        expect(() => pipeline('Article').step('bad', s => s
            .rowNumber('x', { orderBy: 'Id' })
            .derive('X', 'WordCount + 1'))).toThrow(/X/);
    });
});

File: Iverson.Clients/Go/iverson_test/pipeline_test.go
package iverson_test

import (
	"testing"

	pb "github.com/iverson/clients/go/generated"
	"github.com/iverson/clients/go/iverson"
)

func numberVal(n float64) *pb.SearchValue {
	return &pb.SearchValue{Kind: &pb.SearchValue_NumberVal{NumberVal: n}}
}

func boolVal(b bool) *pb.SearchValue {
	return &pb.SearchValue{Kind: &pb.SearchValue_BoolVal{BoolVal: b}}
}

func TestPipelineBuildFull(t *testing.T) {
	req, err := iverson.NewPipeline("Article").
		Where("IsPublished", pb.SearchOperator_EQUALS, boolVal(true)).
		Step("by_author", func(s *iverson.PipelineStepBuilder) {
			s.GroupBy("AuthorId").
				CountAll("articles").
				Having("articles", pb.SearchOperator_GREATER_THAN, numberVal(5))
		}).
		Step("ranked", func(s *iverson.PipelineStepBuilder) {
			s.RowNumber("rank", "", "articles", true)
		}).
		Step("named", func(s *iverson.PipelineStepBuilder) {
			s.Join("Author", "AuthorId", "Id").
				SelectAllFrom("ranked").
				SelectPick("Author", "Name", "author_name")
		}).
		SortOnDesc("rank").
		Limit(5).
		Build()

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if req.TypeName != "Article" || len(req.Steps) != 3 || req.Limit != 5 {
		t.Fatalf("unexpected request shape: %+v", req)
	}
	if req.Steps[0].GroupBy[0].Field != "AuthorId" {
		t.Errorf("groupBy field = %q", req.Steps[0].GroupBy[0].Field)
	}
	if req.Steps[1].Windows[0].Kind != pb.WindowFunctionKind_ROW_NUMBER ||
		!req.Steps[1].Windows[0].Descending {
		t.Errorf("window = %+v", req.Steps[1].Windows[0])
	}
	if req.Steps[2].Joins[0].Source != "Author" ||
		req.Steps[2].Joins[0].On[0].Left != "AuthorId" {
		t.Errorf("join = %+v", req.Steps[2].Joins[0])
	}
	if !req.Steps[2].Select[0].All || req.Steps[2].Select[1].Alias != "author_name" {
		t.Errorf("select = %+v", req.Steps[2].Select)
	}
}

func TestPipelineRunningSumWithPartitionBy(t *testing.T) {
	req, err := iverson.NewPipeline("Article").
		Step("cumulative", func(s *iverson.PipelineStepBuilder) {
			s.RunningSum("running_total", "AuthorId", "ViewCount", "PublishedAt")
		}).
		Build()
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	win := req.Steps[0].Windows[0]
	if win.Kind != pb.WindowFunctionKind_RUNNING_SUM {
		t.Errorf("kind = %v", win.Kind)
	}
	if win.PartitionBy != "AuthorId" {
		t.Errorf("partitionBy = %q, want %q", win.PartitionBy, "AuthorId")
	}
	if win.Field != "ViewCount" || win.OrderBy != "PublishedAt" {
		t.Errorf("field/orderBy = %q/%q", win.Field, win.OrderBy)
	}
}

func TestPipelineDuplicateStepNameErrors(t *testing.T) {
	_, err := iverson.NewPipeline("Article").
		Step("x", func(s *iverson.PipelineStepBuilder) { s.Derive("a", "WordCount") }).
		Step("X", func(s *iverson.PipelineStepBuilder) { s.Derive("b", "WordCount") }).
		Build()
	if err == nil {
		t.Fatal("expected duplicate step name error")
	}
}

func TestPipelineReadsUnknownStepErrors(t *testing.T) {
	_, err := iverson.NewPipeline("Article").
		Step("a", func(s *iverson.PipelineStepBuilder) { s.Reads("nope") }).
		Build()
	if err == nil {
		t.Fatal("expected unknown reads error")
	}
}

func TestPipelineWindowAndGroupByErrors(t *testing.T) {
	_, err := iverson.NewPipeline("Article").
		Step("bad", func(s *iverson.PipelineStepBuilder) {
			s.RowNumber("rn", "", "Id", false).GroupBy("AuthorId").CountAll("n")
		}).
		Build()
	if err == nil {
		t.Fatal("expected windows-XOR-aggregation error")
	}
}

func TestPipelineJoinWithoutSelectErrors(t *testing.T) {
	_, err := iverson.NewPipeline("Article").
		Step("bad", func(s *iverson.PipelineStepBuilder) {
			s.Join("Author", "AuthorId", "Id")
		}).
		Build()
	if err == nil {
		t.Fatal("expected join-requires-select error")
	}
}

func TestPipelineDuplicateAliasErrors(t *testing.T) {
	_, err := iverson.NewPipeline("Article").
		Step("bad", func(s *iverson.PipelineStepBuilder) {
			s.RowNumber("x", "", "Id", false).Derive("X", "WordCount + 1")
		}).
		Build()
	if err == nil {
		t.Fatal("expected duplicate alias error")
	}
}

func TestJoinOn_WithCompositeKey_AddsMultipleConditions(t *testing.T) {
	req, err := iverson.NewPipeline("Article").
		Step("enriched", func(s *iverson.PipelineStepBuilder) {
			s.JoinOn("Author", []iverson.JoinCondition{
				{Left: "AuthorId", Right: "Id"},
				{Left: "TenantId", Right: "TenantId"},
			}).
				SelectAllFrom("base")
		}).
		Build()
	if err != nil {
		t.Fatalf("Build: %v", err)
	}
	join := req.Steps[0].Joins[0]
	if len(join.On) != 2 {
		t.Fatalf("expected 2 join conditions, got %d", len(join.On))
	}
	if join.On[0].Left != "AuthorId" || join.On[0].Right != "Id" {
		t.Errorf("unexpected first condition: %+v", join.On[0])
	}
	if join.On[1].Left != "TenantId" || join.On[1].Right != "TenantId" {
		t.Errorf("unexpected second condition: %+v", join.On[1])
	}
}

func TestStep_WithReservedNameBase_ReturnsBuildError(t *testing.T) {
	_, err := iverson.NewPipeline("Article").
		Step("base", func(s *iverson.PipelineStepBuilder) {}).
		Build()
	if err == nil {
		t.Fatal("expected an error for reserved step name \"base\"")
	}
}

func TestStep_WithMetricsButNoGroupBy_ReturnsBuildError(t *testing.T) {
	_, err := iverson.NewPipeline("Article").
		Step("s1", func(s *iverson.PipelineStepBuilder) { s.Count("id") }).
		Build()
	if err == nil {
		t.Fatal("expected an error for metrics without groupBy")
	}
}
